feature-pitches/010-compute-sessions/compute-sessions.md

```diff
@@ -0,0 +1,220 @@
+
+## 🤔 Problem
+
+Renku 1.0 needs sessions! But more than that, we want to improve on the current sessions, solving the following user problems:
+
+- When I am starting a new project, I want to **start out with a basic compute environment** (for example with a pre-built data science stack), and I’ll decide what extra packages I need later.
+- When I create a new Renku project, I want to get started right away with a pre-built environment- **I don’t want to have to wait for the environment image to be built** to get started.
+- When I start a session, I want to **clearly understand what environment image the session will use and what is installed in it**, so I am not confused where the compute environment comes from.
+- When I go to launch a session on a project where I recently made an edit to a code file that does not impact the project environment, **I don’t want to have to wait for the image to re-build** since the file I changed didn’t change the environment anyway.
+
+There are a few ways we want 1.0 Environments to work in the future that we won’t be able to make work right away (these are *out of scope* for this pitch), but I include them here to paint a bigger picture and give some extra context for the direction we want Environments to go.
+
+- When I have already created a Docker image for my project that I use in other places, **I want Renku to use that pre-exising image** so I don’t have to re-create a new environment.
+- When I know my teammate has already created a great environment they do their work in, I want to **search and reuse that environment** in my own project in RenkuLab.
+- When I create a dashboard in a Renku Project, I want to **provide a separate entry point for app viewers** that does not reveal the development environment I use to build the app, so that my app consumers are not confused.
+- When I use my Renku project image in other places, I want the **image to be small** so it’s easy to move to other systems (aka not have Jupyter installed in it if I don’t need it).
+
+## 🍴 Appetite
+
+6 weeks.
+
+## 🎯 Solution
+
+In Renku 1.0 (generally, not already in this pitch!), we want to enable 2 key new features:
+
+- Enabling more than one environment per project
+- Sharing environments between projects
+
+For these features, we need some new terminology.
+
+### Terminology
+
+#### Environment
+
+An environment consists of a docker image plus metadata. An environment is independent of any one project and can be reused in multiple projects.
+
+An environment has the following properties:
+
+- name
+- an image/tag
+- metadata:
+  - who created it
+  - date of creation
+- [optional] description (e.g. why it is useful)
+- [optional] keywords/tags to aid in searching (for example, I might tag an environment with `#python` `#pytorch` to give a quick impression of what is in the image)
+
```

Comment:
In a very general sense, I feel that these changes are centered on the actual Docker images Renku is using. This is very helpful, however there is something that we cannot do with docker images: "mixing" the environments. 

Let's put a concrete case:

Let's say I have an app for "paraview" GUI viewer and an app for a finite element software and another and finally one for Solidipes. The user would wish to create a single session with /paraview, /solidipes, /fe and /lab automatically. In the current Renku, I would do with python plugins and jupyter proxy: this is conditioned by the currently installed python packages. But what is proposed here, do not allow that, I think.

How am I supposed to mix environments ? Can I create as session with two environments ? From your proposal, I think that someone has to make a Dockerfile (somewhere) and describe the installation for the mix.
However, it would make sense to split the process of installation (and not requiring the users to write the Dockerfile for any combination). 

I think the metadata attached to an environment should also include possible (compatible) requirements. That would allow the dependencies to be somehow selected while creating the session (and select them in a catalog). 

Another possibility could be to have two types of "environments": one for the Dockerfile level and one for the requirements. (Sadly, this is orthogonal to the request that users do not want to wait installation times when spawning a new session....).

To summarize: I think it is important to think of a way/procedure, which allows mixing environment/apps easily.

Reply:
yikes, we left this unanswered for a little too long. You are totally right @anciaux that mixing environments should be possible, but I think that has to be done after this initial work (btw, this is already implemented). There are two possible options for what you are proposing:

1. pluggable front-ends (the user provides a docker image with their runtime, but another independent front-end is added to their session)
2. multi-container sessions

I think both of these have use-cases and I would like to see us work out how to support them. In the case of front-ends (option 1), they would need to be configured and maintained by the admins of the platform. The disadvantage is that this would be less flexible for the user, but probably lead to better UX because it would be more controlled and carefully managed. Option 2 gives more flexibility but the containers would not share the runtime so in the case of e.g. python or other packages needed across containers, it would not be straightforward to share them.